prime-router/docs/design/proposals/0024-original-message-passthrough.md

```diff
@@ -0,0 +1,168 @@
+      modified.
+    * *Pro*: There is no limitation on incoming file format.
+    * *Con*: This approach is highly reliant on the processes performed at the `receive` pipeline step remaining as they
+      are now. Any refactoring of this code must take original message passthrough into consideration.
+
+## Limitations
+
+* Filtering can still occur as a routing process, but the nature of sending an unmodified HL7v2 file will mean any
+  conditions that would have been eliminated by filtering would be sent regardless.
+* `MERGE` batching cannot be performed for original passthrough HL7v2 files.
+* Filenames provided by the sender must be unique. New submissions with an existing filename should be rejected.
+* Filenames in the Azure blobstore currently reflect pipeline specific information to more easily identify when and why
+  a file was created.
+    * There is currently a limitation that files must have unique names, and blobURLs cannot be reused.
+    * External filenames are in many cases based on the blobURL presently.
+
+## Feature Design
+
+After performing analysis of the possible implementations above, the **Add feature to the UP translate step: retrieve
+blobURL from report_file DB table** implementation was selected.
+
+The overall design of the original passthrough feature is as follows:
+
+* A new `Topic` will be used for original passthrough items. This guarantees that data that must be handled via original
+  passthrough is not processed as other full ELR data and is only sent to receivers in its original state. This also
+  allows receivers to be able to configure endpoints specific to the data for this particular sender, separate from any
+  batching settings the receiver may otherwise have configured.
+* As the `Topic` is unique, no other identifiers for original passthrough items are needed. Configuring the sender and
+  receivers for the original passthrough Topic is all that is required.
+* The `translate` step will copy the original blob data from the blobstore instead of translating from the internal
+  FHIR bundle. Enhancements to `BlobAccess` that have been made in preparation for the `Manage Translation Schemas`
+  feature can be utilized to perform this operation. The `translate` step will then directly insert the resulting blob
+  to the `send` queue.
+* `Report.formExternalFilename` uses the filename of the blobstore item, so there should not be any further
+  modifications necessary as long as the blobstore item is named correctly by the `translate` function.
+
+### Issues to be created
+
+* Add header to `receive` step to allow sender to supply original filename and store in database
+  field `report_file.external_name`
+    * Submissions that provide an `external_name` that already exists should be rejected.
+* Write function to look up original filename and blob URL for a given report ID
+* Add step to `translate` function to check message topic, copy the original blob with original filename (under receiver
```

Comment:
I think this works, but double check: If a receiver comes back and says "Yo, I don't know the report ID but I got a file named File-ABC-123.hl7 from ya'll and it has these issues with it" do we have enough info to figure out where that came from? What if we sent many files with that name? It seems like we need to append the report ID to the file name, which ELIMS would be okay with, but we need to make sure file name length is still reasonable and won't cause issues.

Reply:
Also, what if the the transport protocol is REST? Do we put the original file name in the header in that case?

Reply:
Unless I'm mistaken, REST transport already includes the filename under `HttpHeaders.ContentDisposition`